### PR TITLE
Compute metrics

### PR DESCRIPTION
**Pull request by jeffreymeetkai**
- `preprocess_logits_for_metrics`  function to reduce CPU memory requirements during evaluation
- accuracy and loss evaluation implementation

### CODE REVIEW

File: train/README.md [hunk 1]
```diff
@@ -34,6 +34,33 @@ torchrun --nproc_per_node=2 --master_port=20001 train.py \
     --model_max_length 4096 \
     --gradient_checkpointing True \
     --output_dir functionary-v1
+
+# 2xA100 80GB WITH evaluation
+torchrun --nproc_per_node=2 --master_port=20001 train.py \
```
Comment (musab-mk) [inline train/README.md:39]: Can we remove the "WITHOUT Evaluation" bash code and only keep the "WITH" eval one for simplicity

File: train/README.md [hunk 2]
```diff
@@ -5,13 +5,13 @@ python3 -m venv venv && source venv/bin/activate
 # Install Torch 2.0.1
 pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu118
 
-# Install Dependencies, (Latest main version of huggingface is critical as its giving OOM without it, We need to use this until 4.32 is out)
-pip install accelerate==0.21.0 git+https://github.com/huggingface/transformers sentencepiece packaging ninja einops wandb
+# Install Dependencies, (huggingface version 4.32 as earlier versions results in OOM)
+pip install accelerate==0.21.0 transformers==4.32.0 sentencepiece packaging ninja einops wandb
```
Comment (musab-mk) [inline train/README.md:9]: Can we also ping the other ones

File: train/train.py [hunk 3]
```diff
@@ -241,6 +244,27 @@ def safe_save_model_for_hf_trainer(trainer: transformers.Trainer, output_dir: st
         cpu_state_dict = {key: value.cpu() for key, value in state_dict.items()}
         del state_dict
         trainer._save(output_dir, state_dict=cpu_state_dict)  # noqa
+        
+        
+def split_data(raw_data, input_file, percentage):
```
Comment (musab-mk) [inline train/train.py:249]: Since this is fully dataset related, Let's separate this into another module

File: train/train.py [hunk 4]
```diff
@@ -284,13 +308,50 @@ def train():
...
-    dataset = CustomDataset(raw_data, tokenizer)
+    if training_args.do_eval:
+        # Take 90:10 train-validation split
+        raw_train_data, raw_eval_data = split_data(raw_data, data_args.data_path, 0.9)
+        train_dataset = CustomDataset(raw_train_data, tokenizer)
+        eval_dataset = CustomDataset(raw_eval_data, tokenizer)
+    else:
+        train_dataset = CustomDataset(raw_data, tokenizer)
+        
+    def preprocess_logits_for_metrics(logits, labels):
```
Comment (musab-mk) [inline train/train.py:319]: Can you type concise docstrin on what this code does? Also please refactor this function into to be much clear to understand (Seperate lines and space breaks etc.)

File: train/train.py [hunk 5]
```diff
@@ -284,13 +308,50 @@ def train():
...
+        shift_logits = logits[..., :-1, :].contiguous()
+        shift_labels = labels[..., 1:].contiguous()
+        loss_fct = CrossEntropyLoss(reduction="none")
+        shift_logits = shift_logits.view(-1, tokenizer.vocab_size)
+        shift_labels = shift_labels.view(-1)
+        loss = loss_fct(shift_logits, shift_labels)
+        loss = torch.mean(loss.view(logits.shape[0], -1), dim=-1)
+        return pred_ids, loss
+    
+    def compute_metrics(eval_preds):
```
Comment (musab-mk) [inline train/train.py:330]: Let's use appropriate blank spaces here too please. Also What is B x L ? Let's clarify them too.

File: train/train.py [hunk 6]
```diff
@@ -284,13 +308,50 @@ def train():
...
 
     trainer = Trainer(
         model=model,
         tokenizer=tokenizer,
         args=training_args,
-        train_dataset=dataset,
+        train_dataset=train_dataset,
+        eval_dataset=eval_dataset if training_args.do_eval else None,
+        compute_metrics=compute_metrics if training_args.do_eval else None,
+        preprocess_logits_for_metrics=preprocess_logits_for_metrics if training_args.do_eval else None,
```
Comment (musab-mk) [inline train/train.py:354]: This whole one liner if else statements look complicated. Can you find a better way? If not, we can keep it.

File: train/train.py [hunk 7]
```diff
@@ -284,13 +308,50 @@ def train():
...
+        train_dataset = CustomDataset(raw_train_data, tokenizer)
+        eval_dataset = CustomDataset(raw_eval_data, tokenizer)
+    else:
+        train_dataset = CustomDataset(raw_data, tokenizer)
+        
+    def preprocess_logits_for_metrics(logits, labels):
+        pred_ids = torch.argmax(logits, dim=-1)
+        shift_logits = logits[..., :-1, :].contiguous()
+        shift_labels = labels[..., 1:].contiguous()
+        loss_fct = CrossEntropyLoss(reduction="none")
```
Comment (musab-mk) [inline train/train.py:323]: If you mean loss function, can we rename this to "loss_fn" or "loss_function" to keep consistency on the names.

File: train/train.py [hunk 8]
```diff
@@ -284,13 +308,50 @@ def train():
 
     random.shuffle(raw_data)
 
-    dataset = CustomDataset(raw_data, tokenizer)
+    if training_args.do_eval:
+        # Take 90:10 train-validation split
+        raw_train_data, raw_eval_data = split_data(raw_data, data_args.data_path, 0.9)
```
Comment (musab-mk) [inline train/train.py:313]: This dataset split shouldn't be hardcoded. Let's take it with args

**Review by musab-mk — CHANGES_REQUESTED** [review]

File: train/README.md (hunk 2 above)
Comment (musab-mk) [reply train/README.md:9]: Did we try 4.32.1. ? Cause there was a tokenizer fix in that release. Lets use that if its non-problematic

**Review by musab-mk — CHANGES_REQUESTED** [review]

File: train/README.md (hunk 2 above)
Comment (jeffreymeetkai) [reply train/README.md:9]: I did not because the previous comment on line 8 mentioned that 4.32.0 will fix the OOM issue so I just tried 4.32.0 only. Will try 4.32.1

**Review by musab-mk — APPROVED** [review]
